quickwit/quickwit-ingest/src/ingest_v2/ingester.rs

```diff
@@ -0,0 +1,1257 @@
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::iter::once;
+use std::path::Path;
+use std::sync::Arc;
+use std::time::Duration;
+use std::{cmp, fmt};
+
+use async_trait::async_trait;
+use bytes::{Bytes, BytesMut};
+use futures::stream::FuturesUnordered;
+use futures::StreamExt;
+use mrecordlog::MultiRecordLog;
+use quickwit_common::tower::Pool;
+use quickwit_common::ServiceStream;
+use quickwit_proto::ingest::ingester::{
+    AckReplicationMessage, FetchResponseV2, IngesterService, IngesterServiceClient,
+    IngesterServiceStream, OpenFetchStreamRequest, OpenReplicationStreamRequest,
+    OpenReplicationStreamResponse, PersistRequest, PersistResponse, PersistSuccess, PingRequest,
+    PingResponse, ReplicateRequest, ReplicateSubrequest, SynReplicationMessage, TruncateRequest,
+    TruncateResponse, TruncateSubrequest,
+};
+use quickwit_proto::ingest::{CommitTypeV2, IngestV2Error, IngestV2Result, ShardState};
+use quickwit_proto::types::{NodeId, QueueId};
+use tokio::sync::{watch, RwLock};
+
+use super::fetch::FetchTask;
+use super::replication::{
+    ReplicationClient, ReplicationClientTask, ReplicationTask, ReplicationTaskHandle,
+};
+use super::IngesterPool;
+use crate::metrics::INGEST_METRICS;
+use crate::DocCommand;
+
+#[derive(Clone)]
+pub struct Ingester {
+    self_node_id: NodeId,
+    ingester_pool: IngesterPool,
+    mrecordlog: Arc<RwLock<MultiRecordLog>>,
+    state: Arc<RwLock<IngesterState>>,
+    replication_factor: usize,
+}
+
+impl fmt::Debug for Ingester {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Ingester")
+            .field("replication_factor", &self.replication_factor)
+            .finish()
+    }
+}
+
+pub(super) struct IngesterState {
+    pub primary_shards: HashMap<QueueId, PrimaryShard>,
+    pub replica_shards: HashMap<QueueId, ReplicaShard>,
+    pub replication_clients: HashMap<NodeId, ReplicationClient>,
+    pub replication_tasks: HashMap<NodeId, ReplicationTaskHandle>,
+}
+
+impl IngesterState {
+    fn find_shard_status_rx(&self, queue_id: &QueueId) -> Option<watch::Receiver<ShardStatus>> {
+        self.primary_shards
+            .get(queue_id)
+            .map(|shard| shard.shard_status_rx.clone())
+            .or_else(|| {
+                self.replica_shards
+                    .get(queue_id)
+                    .map(|shard| shard.shard_status_rx.clone())
+            })
```

Comment:
nitpick. Whiel prosaic, I sincerely think the following is more readable.

```suggestion
        if let Some(shard) = self.primary_shards.get(queue_id) {
             return Some(shard.shard_status_rx.clone());
        }
        if let Some(shard) = self.replica_shards.get(queue_id) {
            return Some(shard.shard_status_rx.clone());
        }
        None
```